iwyu.cc

```diff
@@ -1650,13 +1676,34 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       return;
 
     // Figure out the best location to attribute uses inside macros.
-    if (IsInMacro(used_loc))
-      used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
     const FileEntry* used_in = GetFileEntry(used_loc);
+    const NamedDecl* fwd_decl = NULL;
+    bool isMacroDefinedAndUsedInSameFile = false;
+    if (IsInMacro(used_loc)) {
+      isMacroDefinedAndUsedInSameFile =
+          (GetFileEntry(used_loc) ==
+           GetFileEntry(GetInstantiationLoc(used_loc)));
+
+      const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, used_loc);
+      if (fwd_decl) {
+        // Make sure we keep the heuristic forward decl.
+        preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
+            used_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
+            comment);
+
+        // Attribute use to caller
+        used_loc = GetInstantiationLoc(used_loc);
+      } else {
+        used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
+      }
+    }
 
-    preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
-        used_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
-        comment);
+    used_in = GetFileEntry(used_loc);
+    if (isMacroDefinedAndUsedInSameFile || !fwd_decl) {
```

Comment:
It requires more attention to understand that in case of no macro `isMacroDefinedAndUsedInSameFile` will be false and `fwd_decl` will be NULL, so you need to report forward declare use. Maybe have this condition inside `IsInMacro` block and maybe it will be easier to understand when condition is telling when you want to avoid `ReportForwardDeclareUse`.

Another issue, don't know how important it is. `fwd_decl` is a pointer but is used as a boolean variable. Will it be better to have it as a boolean variable?